text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,590 @@
+e. g. it should be to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures, that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be attached to
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
+the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
+"Receive" span.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
+MAY link to the "Create" or "Publish" span for the message.
+
+## Examples
+
+This section contains a list of examples illustrating the use of the
+conventions outlined above. Green boxes denote spans that are required to
+exist in order to conform to those conventions. Other boxes denote spans that
+are not required and covered by the conventions, but are hopefully helpful in
+understanding how messaging spans can be integrated into an overall trace flow.
+
+### Single message producer, single message push-based consumer
```

Comment:
Do you think it would make sense in separating the diagrams in "push-based" and "pull-based" sub-sections? Right now 1 they are mixed which makes it hard to find cases and 2 we repeat "push/pull-based" in all diagrams, which makes the diagram title long.

Reply:
I did so, also replaced headers for individual cases with a more verbose description.